File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    static void findCombination(int i,int[] candidates,int target,List<List<Integer>> result, List<Integer> combination){
        if(i == candidates.length || target == 0){
            if(target == 0){
                result.add(new ArrayList<>(combination));
            }
            return;
        }

        if(candidates[i] <= target){
            combination.add(candidates[i]);
            findCombination(i, candidates, target-candidates[i], result,combination);
            combination.remove(combination.size() - 1);
        }

        findCombination(i+1, candidates, target, result,combination);        
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List < List < Integer >> result = new ArrayList < > ();
        findCombination(0, candidates, target, result, new ArrayList < > ());
        return result;
    }
}

File: 0039-combination-sum/0039-combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    candidates.sort((a, b) => a - b);
    var result = [];
    generateCombinationSum(result,[],0,candidates,target);
    return result;
};

function generateCombinationSum(result,current,i,candidates,target){
    if(i == candidates.length || target == 0){
        if(target == 0){
            result.push([...current]);
        }
        return;
    }

    if(target >= candidates[i]){
        current.push(candidates[i]);
        generateCombinationSum(result,current,i,candidates,target-candidates[i]);
        current.pop();
    }
    generateCombinationSum(result,current,i+1,candidates,target);

}